### PR TITLE
Adding an example on Conditional GANs

### PR DESCRIPTION
**Pull request by sayakpaul**
This example demonstrates the workflow of a basic yet powerful Conditional GAN. This is an important element for a majority of image generation architectures like VQ-GANs, DALL-E, etc. 

A Colab Notebook is available [here](https://colab.research.google.com/drive/1E0Pgp2HKQYJreDBb-Zx88TQLxwArBEjw?usp=sharing).

### CODE REVIEW

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This is very high-quality code, as usual 👍 

You appear to have two files, one at `conditional_gans.py` (delete this one) and the other at `examples/vision/conditional_gans.py`.

I recommend renaming to `conditional_gan.py` (same in the title), to be consistent with our other GAN examples.

I'll follow up with copyedits afterwards.

**Comment by sayakpaul** [comment]
Appreciate the kind words, @fchollet. Part of the code comes from your amazing guide: [Customizing what happens in fit()](https://keras.io/guides/customizing_what_happens_in_fit/). I have ensured the citation in the text as well. 

Also, I noticed that we should have this example under `generative` section so I moved it accordingly. Let me know your thoughts.

File: examples/generative/conditional_gan.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+"""
+"""
+Generative Adversarial Networks (GANs) let us generate novel samples from noise, adapt
+one domain into another, and so on. For the first case, the noise is generally sampled
+from a normal distribution and then it goes through a series of transformations that turn
+the noise into something plausible (image, video, audio, etc.). Now, if we wanted
+to generate novel hand-written digits using a GAN. But we also wanted to *condition* its
+outputs on the class of digit we want to generate i.e. zero, one, two, three, and so on.
+[Vanilla DCGAN](https://arxiv.org/abs/1511.06434) won't allow us to achieve this because
+it does have that conditional aspect in its setup.
```
Comment (s-mrb) [inline examples/generative/conditional_gan.py:16]:
```suggestion
the noise into something plausible (image, video, audio, etc.). Now, if we want
to generate novel hand-written digits using a GAN, and also want to *condition* its
outputs on the class of digit we want to generate, i.e. zero, one, two, three, and so on, then
[Vanilla DCGAN](https://arxiv.org/abs/1511.06434) won't allow us to achieve this because
it does have that conditional aspect in its setup.
```

File: examples/generative/conditional_gan.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+print(f"Shape of training images: {all_digits.shape}")
+print(f"Shape of training labels: {all_labels.shape}")
+
+"""
+## Calculating the channel dimensions for our GAN
+
+In the unconditional GANs, we first usually start by sampling noise (of some fixed
+dimension) from a normal distribution. But in this case, we also need to need to account
+for the class labels. To do this, we will add the number of classes to:
```
Comment (s-mrb) [inline examples/generative/conditional_gan.py:99]:
```suggestion
dimension) from a normal distribution. But in this case, we also need to account
for the class labels. To do this, we will add the number of classes to:
```

**Review by s-mrb — CHANGES_REQUESTED** [review]

**Comment by sayakpaul** [comment]
@s-mrb thanks for providing the suggestions. That said, I think the text pieces (for which you provided the suggestions) are consistent with what I wanted to convey to the readers. So, I'd just go with them.

**Comment by s-mrb** [comment]
Below sentence seems incomplete:
> Now, if we wanted to generate novel hand-written digits using a GAN`.` 

And I thought you might mistakenly added double need in below sentence:
> But in this case, we also `need` to `need` to account for the class labels.

Have a look. Thanks :)

**Comment by sayakpaul** [comment]
Thanks much for the help. I am sorry that I did not look closely. Appreciate it :)

**Comment by s-mrb** [comment]
> Thanks much for the help. I am sorry that I did not look closely. Appreciate it :)

haha, don't be sorry, we all do this, it is so easy to overlook stuff in opensource work.

**Comment by fchollet** [comment]
Everything looks good! Note that I pushed some copyedits. I think we're ready to merge the generated files.

**Comment by sayakpaul** [comment]
@fchollet all done. Ready for you to take over.

**Review by fchollet — APPROVED** [review]

**Comment by fchollet** [comment]
Thanks for the great contribution! 👍
